### PR TITLE
sql: disallow cross-database sequence references

### LINKED ISSUE #69992: sql: disallow cross-database sequence references
**Is your feature request related to a problem? Please describe.**
https://github.com/cockroachdb/cockroach/issues/54126 and its resolving PR related to sequences https://github.com/cockroachdb/cockroach/pull/54531 we only looked at sequence ownership and not sequence use. This is unfortunate as we want to disallow sequence references altogether.

**Describe the solution you'd like**
Gate cross-database sequence references like we do sequence owners

### PR DESCRIPTION
**Pull request by fqazi**
Fixes: #69992

Previously, we incorrectly allowed cross DB references
for sequences. This was inadequate because we are going
to drop support for cross DB references. To address this
this patch will disable support for cross DB sequence
references by default.

Release note (sql change): Disallow cross DB references for sequences by
default. This can be enabled with the cluster setting
sql.cross_db_sequence_references.enabled.

### CODE REVIEW

**Review by ajwerner — COMMENTED** [review]
> Reviewed 3 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70581)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/rename_table.go, line 337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-MkNHEJHD10D3zt4R1x8:-MkNHEJHD10D3zt4R1x9:b-ttkmno) ([raw file](https://github.com/cockroachdb/cockroach/blob/2820a8222a71884f131039b4fdbf7a832c236aa6/pkg/sql/rename_table.go#L337)):*
> ```Go
> 	// lead to a cross DB reference, and an appropriate
> 	// error is generated.
> 	checkDepForCrossDbRef := func(depID descpb.ID, isOwner bool) error {
> ```

my preference is that you create some consts like:

```go
type crossDBDepType int
const owner, reference crossDBDepType = 0, 1
```
or something like that. 
___
*[pkg/sql/rename_table.go, line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-MkNH3Qn-rmQFtush09o:-MkNH3Qn-rmQFtush09p:bnoiryw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2820a8222a71884f131039b4fdbf7a832c236aa6/pkg/sql/rename_table.go#L353)):*
> ```
> 		// For tables return an error based on if we are depending
> 		// on a view or sequence.
> 		if tableDesc.IsTable() {
> ```
this cascade of conditionals is pretty hard to follow. Can we rework the conditions, perhaps in a switch or nested switches, to be clearer?

**Review by ajwerner — COMMENTED** [review]
Seems like this change broke a test:

```
=== CONT  TestLogic/local/rename_table/rename_table_using_types_to_different_database_not_allowed_55709
    logic.go:1867: 
         
        testdata/logic_test/rename_table:432: 
        expected "this sequence will be OWNED BY a table \"t1s1ref\" in a different database after rename", but no error occurred
--- done: testdata/logic_test/rename_table with config local: 114 tests, 1 failures
```

> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-:-MkgVs8f772FDU-0UGtY:blvbujt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/rename_table.go, line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-MkgWs5m9T5yQRAS4B9m:-MkgWs5m9T5yQRAS4B9n:b18z7h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/da4e446889487c424cdbd6e4574a33cc76659cc2/pkg/sql/rename_table.go#L413)):*
<details><summary><i alt="summary">Quoted 11 lines of code…</i></summary>

> ```
> 		case tableDesc.IsSequence() && depType == reference:
> 			// For sequences dependent objects can only be
> 			// a relations.
> 			return errors.WithHintf(
> 				pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
> 					"this sequence will be referenced BY a table %q in a different database after rename "+
> 						"(see the '%s' cluster setting)",
> 					dependentObject.GetName(),
> 					allowCrossDatabaseSeqReferencesSetting),
> 				crossDBReferenceDeprecationHint(),
> 			)
> ```
</details>

I think you lost the case for `this sequence will be OWNED BY`. 
___
*[pkg/sql/rename_table.go, line 464 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-MkgW2dp5u_0Ma3-W-ud:-MkgW2dp5u_0Ma3-W-ue:bw5hhcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/da4e446889487c424cdbd6e4574a33cc76659cc2/pkg/sql/rename_table.go#L464)):*
> ```Go
> 		// check if any column owns a sequence.
> 		for _, columnDesc := range tableDesc.Columns {
> 			if !allowCrossDatabaseSeqOwner.Get(&p.execCfg.Settings.SV) {
> ```

this two-level setting checking feels like overkill. You already check the setting in `checkDepForCrossDbRef`, let that be the only check. 
___
*[pkg/sql/rename_table.go, line 467 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-MkgVgugCL2lT71SuALH:-MkgVgugCL2lT71SuALI:b-wq5s7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/da4e446889487c424cdbd6e4574a33cc76659cc2/pkg/sql/rename_table.go#L467)):*
> ```
> 					err := checkDepForCrossDbRef(ownsSequenceID, owner)
> 					if err != nil {
> ```
uber-nit: you can inline the error definitions of all of these checks.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70581)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/rename_table.go, line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-MkgWs5m9T5yQRAS4B9m:-Mkgayh9Eu-EjikoP1dG:b-kso3du) ([raw file](https://github.com/cockroachdb/cockroach/blob/da4e446889487c424cdbd6e4574a33cc76659cc2/pkg/sql/rename_table.go#L413)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		case tableDesc.IsSequence() && depType == reference:
> 			// For sequences dependent objects can only be
> 			// a relations.
> 			return errors.WithHintf(
> 				pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
> 					"this sequence will be referenced BY a table %q in a different database after rename "+
> 						"(see the '%s' cluster setting)",
> 					dependentObject.GetName(),
> 					allowCrossDatabaseSeqReferencesSetting),
> 				crossDBReferenceDeprecationHint(),
> 			)
> ```
I think you lost the case for `this sequence will be OWNED BY`. 
</blockquote></details>

Yes, it was accidentally dropped from the switch, can't believe I missed that :man_facepalming:

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70581)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

___
*[pkg/sql/rename_table.go, line 475 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-MkgrzZr-CdFKd7PmlIb:-MkgrzZs-aLGiAq1wq8W:bxj7176) ([raw file](https://github.com/cockroachdb/cockroach/blob/78638fa57b2003d4845787ab39d1040962d87559/pkg/sql/rename_table.go#L475)):*
> ```Go
> 		// check if any column owns a sequence.
> 		for _, columnDesc := range tableDesc.Columns {
> 			if !allowCrossDatabaseSeqOwner.Get(&p.execCfg.Settings.SV) {
> ```

I'm not super fond of the mixing of levels where we check the settings.

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70581)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/rename_table.go, line 475 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-MkgrzZr-CdFKd7PmlIb:-MkhE9RiAWDyXfj_BFk2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78638fa57b2003d4845787ab39d1040962d87559/pkg/sql/rename_table.go#L475)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm not super fond of the mixing of levels where we check the settings. 
</blockquote></details>

Done.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70581#-:-MkhQjfyCRE-p44oaadV:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

**Comment by fqazi** [comment]
@ajwerner TFTR!

bors r+

**Comment by rafiss** [comment]
@fqazi @ajwerner friendly ping - does this still need to be backported?
